osmomath/decimal_test.go

```diff
@@ -30,18 +30,105 @@ func TestDecimalTestSuite(t *testing.T) {
+		return osmomath.PrecisionBigDec
+	case osmomath.Dec:
+		return osmomath.PrecisionDec
+	case osmomath.Int, osmomath.Uint:
+		return 0
+	default:
+		panic(fmt.Sprintf("%T type is not a decimal", decimal))
+	}
+}
+
+// mutResultToAssert is a struct with numerical values scaled to the same precision (0)
+type mutResultToAssert struct {
+	expectedResult        big.Int
+	startValue            big.Int
+	mutativeResult        big.Int
+	nonMutativeResult     big.Int
+	mutativeStartValue    big.Int
+	nonMutativeStartValue big.Int
+}
+
 // assertMutResult given expected value after applying a math operation, a start value,
 // mutative and non mutative results with start values, asserts that mutation are only applied
 // to the mutative versions. Also, asserts that both results match the expected value.
-func (s *decimalTestSuite) assertMutResult(expectedResult, startValue, mutativeResult, nonMutativeResult, mutativeStartValue, nonMutativeStartValue osmomath.BigDec) {
+func (s *decimalTestSuite) assertMutResult(expectedResult, startValue, mutativeResult, nonMutativeResult, mutativeStartValue, nonMutativeStartValue interface{}) {
+	r := scale(expectedResult, startValue, mutativeResult, nonMutativeResult, mutativeStartValue, nonMutativeStartValue)
+
 	// assert both results are as expected.
-	s.Require().Equal(expectedResult, mutativeResult)
-	s.Require().Equal(expectedResult, nonMutativeResult)
+	s.Require().Equal(r.expectedResult, r.mutativeResult)
+	s.Require().Equal(r.expectedResult, r.nonMutativeResult)
 
 	// assert that mutative method mutated the receiver
-	s.Require().Equal(mutativeStartValue, expectedResult)
+	s.Require().Equal(r.mutativeStartValue, r.expectedResult)
 	// assert that non-mutative method did not mutate the receiver
-	s.Require().Equal(nonMutativeStartValue, startValue)
+	s.Require().Equal(r.nonMutativeStartValue, r.startValue)
+}
+
+// scale scales all assertion values to the same big.Int type
+//
+// BigDec, Dec, Int, etc, are just wrappers around big.Int values, some of which are multiplied
+// by some constant factor and are being treated as if they are decimal numbers.
+// Because of this multiplication, it is not possible to directly compare different osmomath numericals (Dec with Int, BigDec with Dec, etc),
+// but if we chop the precision of all numericals, we will be able to compare them.
+// scale() does exactly that.
+//
+// Example:
+//
+// Dec{big.Int{1000000000000000000}} v Int{big.Int{1}}         [CONTRACT: treat Int and Uint as 0-precision numericals]
+//
+// Dec internally does not hold decimal values, it holds big.Int scaled with 18 precision numbers
+// To compare the two above, we need to "scale" them to a common precision.
+// To do that, we need to divide underlying big.Int of Dec by 10^PrecisionDec and then compare it with Int's underlying big.Int
+func scale(expectedResult, startValue, mutativeResult, nonMutativeResult, mutativeStartValue, nonMutativeStartValue interface{}) mutResultToAssert {
```

Comment:
note: this approach to compare different types of osmo numericals might look overcomplicated now, but I think it might come in handy if in the future we add more mutative functions with different types for receiver and return values